File: Ruby/associate-invoiced-products-to-account/helpers/zoho-analytics-api.rb
require_relative 'zoho'

class ZohoAnalyticsApi < Zoho
  attr_accessor :conn

  def initialize
    access_token = self.refresh_access_tokens(ENV['ANALYTICS_REFRESH_TOKEN'])
    self.conn = Faraday.new 'https://analyticsapi.zoho.com/restapi/v2/' do |builder|
      builder.request :authorization , 'Zoho-oauthtoken', -> {access_token}
      builder.request :json
      builder.response :json
      builder.response :raise_error
      builder.response :logger
    end
    @conn.headers['ZANALYTICS-ORGID'] = ENV['ANALYTICS_ORG_ID']
    self
  end

  def export(view, type)
    config = {'CONFIG': {'responseFormat': type}}
    response = self.api_call("workspaces/#{ENV['ANALYTICS_WORKSPACE_ID']}/views/#{view}/data", 'get', config)
    response
  end

  def create_export(view, type)
    config = {'CONFIG': {'responseFormat': type}}
    response = self.api_call("bulk/workspaces/#{ENV['ANALYTICS_WORKSPACE_ID']}/views/#{view}/data", 'get', config)
    return response['status'] == 'success' ? response : 'fail'
  end

  def get_export_details(job_id)
    response = self.api_call("bulk/workspaces/#{ENV['ANALYTICS_WORKSPACE_ID']}/exportjobs/#{job_id}", 'get')
    return response
  end

  def download_exported_data(job_id)
    response = self.api_call("bulk/workspaces/#{ENV['ANALYTICS_WORKSPACE_ID']}/exportjobs/#{job_id}/data", 'get')
    return response
  end
end

File: associate-invoiced-products-to-account/functions/zoho-analytics-api.js
const axios = require('axios');
const dotenv = require('dotenv');

dotenv.config();

class ZohoAnalyticsApi {
    constructor() {
        this.access_token = this.refreshAccessTokens(process.env.ANALYTICS_REFRESH_TOKEN);
        this.conn = axios.create({
            baseURL: 'https://analyticsapi.zoho.com/restapi/v2/',
            headers: {
                'Authorization': `Zoho-oauthtoken ${this.access_token}`,
                'ZANALYTICS-ORGID': process.env.ANALYTICS_ORG_ID
            }
        });
    }

    async refreshAccessTokens(refreshToken) {
        try {
            const response = await axios.post('https://accounts.zoho.com/oauth/v2/token', null, {
                params: {
                    refresh_token: refreshToken,
                    client_id: process.env.CLIENT_ID,
                    client_secret: process.env.CLIENT_SECRET,
                    grant_type: 'refresh_token'
                }
            });
            if (response.data.access_token) {
                return response.data.access_token;
            } else {
                throw new Error('Failed to get access token');
            }
        } catch (error) {
            console.error('Error refreshing access token:', error);
            throw error;
        }
    }

    async apiCall(endpoint, method, config = {}) {
        try {
            const response = await this.conn({
                url: endpoint,
                method: method,
                ...config
            });
            return response.data;
        } catch (error) {
            console.error(`Error making API call to ${endpoint}:`, error);
            throw error;
        }
    }

    async export(view, type) {
        const config = {
            params: {
                'CONFIG': {
                    'responseFormat': type
                }
            }
        };
        return await this.apiCall(`workspaces/${process.env.ANALYTICS_WORKSPACE_ID}/views/${view}/data`, 'get', config);
    }

    async createExport(view, type) {
        const config = {
            params: {
                'CONFIG': {
                    'responseFormat': type
                }
            }
        };
        const response = await this.apiCall(`bulk/workspaces/${process.env.ANALYTICS_WORKSPACE_ID}/views/${view}/data`, 'get', config);
        return response.status === 'success' ? response : 'fail';
    }

    async getExportDetails(jobId) {
        return await this.apiCall(`bulk/workspaces/${process.env.ANALYTICS_WORKSPACE_ID}/exportjobs/${jobId}`, 'get');
    }

    async downloadExportedData(jobId) {
        return await this.apiCall(`bulk/workspaces/${process.env.ANALYTICS_WORKSPACE_ID}/exportjobs/${jobId}/data`, 'get');
    }
}

module.exports = ZohoAnalyticsApi;
